flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/casting/BinaryToBinaryCastRule.java

```diff
@@ -60,18 +62,35 @@ public String generateExpression(
         int inputLength = LogicalTypeChecks.getLength(inputLogicalType);
         int targetLength = LogicalTypeChecks.getLength(targetLogicalType);
 
-        if (context.legacyBehaviour()) {
+        if (context.legacyBehaviour()
+                || ((!couldTrim(targetLength)
+                                // Assume input length is respected by the source
+                                || (inputLength <= targetLength))
```

Comment:
simplify the condition maybe with local variables? or separate branches for early out.

Reply:
```
if (condition) {
  return inputTerm;
}

if (condition) {
  return inputTerm;
}


remaining code without nesting
```

Reply:
I cannot do that apart from the check for the `legacyBehaviour`, the other condition cannot be splitted, it has to be evaluated as a whole.

Reply:
I could revert it maybe to avoid the wrapping `not` but imho it's more readable to exit early or then apply the logic for trimming/padding.